File: dotnet/utils/Xray.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace QAItest.Playwright.DotNet.Utils;

public static partial class Xray
{
    [GeneratedRegex(@"\b[A-Z][A-Z0-9_-]*-\d+\b")]
    private static partial Regex KeyRegex();

    [GeneratedRegex(@"[^A-Za-z0-9._\-\[\] ]+")]
    private static partial Regex SafeCharsRegex();

    public static string SafeName(string value)
    {
        var normalized = value.Replace(Path.DirectorySeparatorChar, '-').Replace(Path.AltDirectorySeparatorChar, '-');
        normalized = SafeCharsRegex().Replace(normalized, "-").Trim(' ', '.');
        return string.IsNullOrWhiteSpace(normalized) ? "test" : normalized;
    }

    public static string NormalizeLabel(string value)
    {
        var normalized = (value ?? string.Empty).Trim().ToLowerInvariant();
        if (string.IsNullOrWhiteSpace(normalized) || normalized.StartsWith("test_", StringComparison.Ordinal))
        {
            return string.Empty;
        }

        return Regex.Replace(normalized, @"[^a-z0-9\-_.]", "-");
    }

    public static List<Dictionary<string, object>> LoadCucumber(string path)
    {
        if (!File.Exists(path))
        {
            return [];
        }

        return JsonSerializer.Deserialize<List<Dictionary<string, object>>>(File.ReadAllText(path)) ?? [];
    }

    public static Dictionary<string, List<string>> SummarizeFailedScenarios(string cucumberFile)
    {
        var output = new Dictionary<string, List<string>>();
        foreach (var feature in LoadCucumber(cucumberFile))
        {
            foreach (var element in AsObjectList(feature, "elements"))
            {
                var name = AsString(element, "name").Trim();
                if (string.IsNullOrWhiteSpace(name))
                {
                    continue;
                }

                var failed = AsObjectList(element, "steps")
                    .Select(step => AsNestedString(step, "result", "status").ToLowerInvariant())
                    .Contains("failed");
                if (!failed)
                {
                    continue;
                }

                var keys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                foreach (var tag in AsObjectList(element, "tags"))
                {
                    var raw = AsString(tag, "name").TrimStart('@');
                    if (KeyRegex().IsMatch(raw))
                    {
                        keys.Add(raw);
                    }
                    else if (raw.StartsWith("TEST_", StringComparison.Ordinal))
                    {
                        var candidate = raw["TEST_".Length..];
                        if (KeyRegex().IsMatch(candidate))
                        {
                            keys.Add(candidate);
                        }
                    }
                }

                if (keys.Count > 0)
                {
                    output[name] = keys.OrderBy(item => item).ToList();
                }
            }
        }

        return output;
    }

    public static Dictionary<string, object> CollectFeatureInfo(string featuresRoot)
    {
        var labels = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var featureFiles = new List<string>();
        if (Directory.Exists(featuresRoot))
        {
            foreach (var path in Directory.EnumerateFiles(featuresRoot, "*.feature", SearchOption.AllDirectories).Order())
            {
                featureFiles.Add(path);
                foreach (var line in File.ReadLines(path))
                {
                    var trimmed = line.Trim();
                    if (!trimmed.StartsWith("@", StringComparison.Ordinal))
                    {
                        continue;
                    }

                    foreach (var token in trimmed.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries))
                    {
                        if (!token.StartsWith("@", StringComparison.Ordinal))
                        {
                            continue;
                        }

                        var normalized = NormalizeLabel(token[1..]);
                        if (!string.IsNullOrWhiteSpace(normalized))
                        {
                            labels.Add(normalized);
                        }
                    }
                }
            }
        }

        return new Dictionary<string, object>
        {
            ["labels"] = labels.OrderBy(item => item).ToList(),
            ["feature_files"] = featureFiles
        };
    }

    private static List<Dictionary<string, object>> AsObjectList(Dictionary<string, object> source, string key)
    {
        if (!source.TryGetValue(key, out var value))
        {
            return [];
        }

        return value switch
        {
            JsonElement element when element.ValueKind == JsonValueKind.Array => element
                .EnumerateArray()
                .Where(item => item.ValueKind == JsonValueKind.Object)
                .Select(item => JsonSerializer.Deserialize<Dictionary<string, object>>(item.GetRawText()) ?? new Dictionary<string, object>())
                .ToList(),
            IEnumerable<Dictionary<string, object>> list => list.ToList(),
            _ => []
        };
    }

    private static string AsString(Dictionary<string, object> source, string key)
    {
        if (!source.TryGetValue(key, out var value) || value is null)
        {
            return string.Empty;
        }

        return value is JsonElement element
            ? element.ValueKind == JsonValueKind.String ? element.GetString() ?? string.Empty : element.ToString()
            : value.ToString() ?? string.Empty;
    }

    private static string AsNestedString(Dictionary<string, object> source, string nestedKey, string finalKey)
    {
        if (!source.TryGetValue(nestedKey, out var nested) || nested is null)
        {
            return string.Empty;
        }

        if (nested is JsonElement element && element.ValueKind == JsonValueKind.Object)
        {
            var map = JsonSerializer.Deserialize<Dictionary<string, object>>(element.GetRawText()) ?? new Dictionary<string, object>();
            return AsString(map, finalKey);
        }

        return nested is Dictionary<string, object> mapValue ? AsString(mapValue, finalKey) : string.Empty;
    }
}

File: java/utils/Xray.java
package utils;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

public final class Xray {
  private static final ObjectMapper MAPPER = new ObjectMapper();
  private static final TypeReference<List<Map<String, Object>>> FEATURE_LIST = new TypeReference<>() {};
  private static final Pattern KEY_RE = Pattern.compile("\\b[A-Z][A-Z0-9_-]*-\\d+\\b");
  private static final Pattern SAFE_CHARS_RE = Pattern.compile("[^A-Za-z0-9._\\-\\[\\] ]+");

  private Xray() {}

  public static String safeName(String value) {
    String normalized = value.replace("\\", "-").replace("/", "-");
    normalized = SAFE_CHARS_RE.matcher(normalized).replaceAll("-").trim();
    normalized = normalized.replaceAll("^[ .]+|[ .]+$", "");
    return normalized.isBlank() ? "test" : normalized;
  }

  public static String normalizeLabel(String value) {
    String normalized = value == null ? "" : value.trim().toLowerCase();
    if (normalized.isBlank() || normalized.startsWith("test_")) {
      return "";
    }
    return normalized.replaceAll("[^a-z0-9\\-_.]", "-");
  }

  public static List<Map<String, Object>> loadCucumber(Path path) throws IOException {
    if (!Files.exists(path)) {
      return List.of();
    }
    return MAPPER.readValue(Files.readString(path), FEATURE_LIST);
  }

  public static Map<String, List<String>> summarizeFailedScenarios(Path cucumberFile) throws IOException {
    Map<String, List<String>> output = new LinkedHashMap<>();
    for (Map<String, Object> feature : loadCucumber(cucumberFile)) {
      for (Map<String, Object> element : asList(feature.get("elements"))) {
        String name = asString(element.get("name")).trim();
        if (name.isBlank()) {
          continue;
        }

        boolean failed = false;
        for (Map<String, Object> step : asList(element.get("steps"))) {
          String status = asString(asMap(step.get("result")).get("status")).toLowerCase();
          if ("failed".equals(status)) {
            failed = true;
            break;
          }
        }
        if (!failed) {
          continue;
        }

        Set<String> keys = new LinkedHashSet<>();
        for (Map<String, Object> tag : asList(element.get("tags"))) {
          String raw = asString(tag.get("name")).replaceFirst("^@", "");
          if (KEY_RE.matcher(raw).matches()) {
            keys.add(raw);
          } else if (raw.startsWith("TEST_")) {
            String candidate = raw.substring("TEST_".length());
            if (KEY_RE.matcher(candidate).matches()) {
              keys.add(candidate);
            }
          }
        }
        if (!keys.isEmpty()) {
          output.put(name, new ArrayList<>(keys));
        }
      }
    }
    return output;
  }

  public static Map<String, Object> collectFeatureInfo(Path featuresRoot) throws IOException {
    List<String> labels = new ArrayList<>();
    List<String> featureFiles = new ArrayList<>();
    if (Files.exists(featuresRoot)) {
      try (var stream = Files.walk(featuresRoot)) {
        stream.filter(path -> path.toString().endsWith(".feature")).sorted().forEach(path -> {
          featureFiles.add(path.toString());
          try {
            String text = Files.readString(path);
            for (String line : text.split("\n")) {
              String trimmed = line.trim();
              if (!trimmed.startsWith("@")) {
                continue;
              }
              for (String token : trimmed.split("\\s+")) {
                if (!token.startsWith("@")) {
                  continue;
                }
                String normalized = normalizeLabel(token.substring(1));
                if (!normalized.isBlank()) {
                  labels.add(normalized);
                }
              }
            }
          } catch (IOException ignored) {
          }
        });
      }
    }

    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("labels", new ArrayList<>(new LinkedHashSet<>(labels)));
    payload.put("feature_files", featureFiles);
    return payload;
  }

  @SuppressWarnings("unchecked")
  private static List<Map<String, Object>> asList(Object value) {
    return value instanceof List<?> list ? (List<Map<String, Object>>) (List<?>) list : List.of();
  }

  @SuppressWarnings("unchecked")
  private static Map<String, Object> asMap(Object value) {
    return value instanceof Map<?, ?> map ? (Map<String, Object>) map : Map.of();
  }

  private static String asString(Object value) {
    return value == null ? "" : String.valueOf(value);
  }
}
